File: typescript/cfg/lang_test.ts
import {Lang} from "wife/typescript/cfg/lang";

let x = new Lang();
x.init_words(`
an indefiniteArticle
a indefiniteArticle
the definiteArticle
eat verb
on prep
apple noun
table noun
is qVerb
does qVerb
edible type
exist type
red adjective
green adjective
clean adjective
blotchy adjective
dirty adjective
`);
x.init_rules(`
S: Verb NounClause | Verb NounClause PrepClause | QuestionVerb NounClause Type | QuestionVerb NounClause PrepClause Type
NounClause: Count ANoun | Adjective Noun | noun
PrepClause: Prep NounClause
ANoun: Adjective Noun | noun
QuestionVerb: qVerb
Adjective: adjective
Type: type
Prep: prep
Verb: verb
Noun: noun
Count: definiteArticle | indefiniteArticle | number
`);
const sentence = "eat the clean apple on the green table";
console.log(sentence);
console.log("result", x.parse_sentence(sentence));

File: engine/cfg/test/lang-test.cc
#include "lang.hh"

#include <filesystem>
#include <fstream>
#include <sstream>
#include <string>

#include "gtest/gtest.h"

namespace fs = std::filesystem;
namespace {

Str read_file(fs::path path) {
  std::ifstream file{path};
  const auto size = std::filesystem::file_size(path);
  Str result(size, ' ');
  file.read(result.data(), size);
  return result;
}

TEST(Lang, ParseSentence_Exists) {
  auto lang = cfg::Lang();
  auto rules = read_file("engine/cfg/test-lang-rules.txt");
  lang.init_rules(rules);
  auto words = read_file("engine/cfg/test-lang-words.txt");
  lang.init_words(words);
  Str sentence = "eat a clean apple";
  const auto* parsed = lang.parse_sentence(sentence, {"ActionSentence"});
  const auto* ast = parsed->ast;
  std::ostringstream ss;
  ss << *ast;
  delete parsed;
  EXPECT_EQ(
      "\n"
      "(RULE\n"
      "(TAGGED Verb, (WORD verb, eat))\n"
      "(TAGGED NounClause, \n"
      "(RULE\n"
      "(TAGGED Count, (WORD indefiniteArticle, a))\n"
      "(TAGGED ANoun, \n"
      "(RULE\n"
      "(TAGGED Adjective, (WORD adjective, clean))\n"
      "(TAGGED Noun, (WORD noun, apple))\n"
      "))\n"
      "))\n"
      ")",
      ss.str());
}

TEST(Lang, InitsRules) {
  auto lang = cfg::Lang();
  auto rules = read_file("engine/cfg/test-lang-rules.txt");
  lang.init_rules(rules);
}

}  // namespace
